### PR TITLE
add retries and delays to work around ESP32-S3 problems

### PR DESCRIPTION
**Pull request by dhalbert**
ESP32-S3 has some issues with the clock-stretching and sleeping done by the LC709203F. See https://github.com/adafruit/circuitpython/issues/6311.

- Add retries when scanning for the device, and also add some delays during setup. This seems to mostly ameliorate the problems on the S3.
- Raise `ValueError` for bad args, not `AttributeError`.
- Raise `OSError` instead of `RuntimeError` on CRC failure. This makes it easier to catch all read/write failures so you can retry.

Tested with this program. Note the `try`-`except` which catches `OSError`. There can still be occasional failures, so it's good to catch and retry.

```py
import time
import board
import busio
from adafruit_lc709203f import LC709203F, PackSize

uart = board.UART()
battery_monitor = LC709203F(board.I2C())
battery_monitor.pack_size = PackSize.MAH400

while True:
    try:
        print("Battery Percent: {:.2f} %".format(battery_monitor.cell_percent))
        print("Battery Voltage: {:.2f} V".format(battery_monitor.cell_voltage))

        uart.write(b"Battery Percent: {:.2f} %\r\n".format(battery_monitor.cell_percent))
        uart.write(b"Battery Voltage: {:.2f} V\r\n".format(battery_monitor.cell_voltage))
    except OSError:
        print("retrying reads")

    time.sleep(2)
```

@Gingertrout @BeatArnet @mopore @ThomasAtBBTF Could you try this changed version and see if it works for you? Thank you. You just need to copy the `adafruit_lc709203f.py` file in this pull request to your CIRCUITPY drive, which will override what is in your `lib/`. You should be able to remove the `i2c.scan()` code from your existing programs, since this does approximately the same thing.

New code is available at https://github.com/adafruit/Adafruit_CircuitPython_LC709203F/blob/1007e4c96bc465204de0cb5a3684022e9e0fb71b/adafruit_lc709203f.py

### CODE REVIEW

**Comment by BeatArnet** [comment]
After downloading the latest version of the adafruit_lc709203f.mpy library and running your sample code on my feather_s3_tft I got the following errors:

>>> %Run -c $EDITOR_CONTENT
Zurückverfolgung (jüngste Aufforderung zuletzt):
  Datei "<stdin>", Zeile 7, in <module>
  Datei "adafruit_lc709203f.py", Zeile 132, in __init__
  Datei "adafruit_lc709203f.py", Zeile 136, in init_RSOC
  Datei "adafruit_lc709203f.py", Zeile 257, in _write_word
OSError: [Errno 116] ETIMEDOUT
>>>

**Comment by dhalbert** [comment]
@BeatArnet The proposed fix is not yet available as .mpy. Please try:
https://github.com/adafruit/Adafruit_CircuitPython_LC709203F/blob/1007e4c96bc465204de0cb5a3684022e9e0fb71b/adafruit_lc709203f.py

**Comment by BeatArnet** [comment]
Unfortunately, after some impressive results…
```
Battery Percent: 100.00 %

Battery Voltage: 4.19 V

Battery Percent: 100.00 %

Battery Voltage: 4.19 V

Battery Percent: 100.00 %

Battery Voltage: 4.19 V

Battery Percent: 100.00 %

Battery Voltage: 4.19 V

Battery Percent: 100.00 %

Battery Voltage: 4.19 V

Battery Percent: 100.00 %

Battery Voltage: 4.19 V

Battery Percent: 100.00 %

Battery Voltage: 4.19 V

Battery Percent: 100.00 %

Battery Voltage: 4.19 V

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads

retrying reads
```

**Comment by dhalbert** [comment]
@BeatArnet hmm, I saw a few missing reads, but not continuous as you did :slightly_frowning_face: . Are you using exactly the same test program, and which board are you testing on?

**Comment by BeatArnet** [comment]
I used your sample-code and the latest library *.py. I have an Adafruit S3 TFT board.

**Review by tekktrik — APPROVED** [review]
All makes sense to me!  Question of of curiosity, why is `ValueError` what's being raised by the failed instancing of `I2CDevice` and not something like `OSError` itself?

**Review by tekktrik — CHANGES_REQUESTED** [review]
Ah, I missed the ongoing conversation.  Requesting changes so it doesn't show as "approved" until that's sorted out, my bad!

**Review by FoamyGuy — APPROVED** [review]
These changes look good to me. I think it's worthwhile to go ahead and merge this improvement even if there is still some situations that can still be problematic for other reasons. 

I tested the code posted in this PR on a Feather S3 TFT with 8.0.0.beta.5. It has run successfully for several minutes for me with the branch from this PR.

@dhalbert or @tekktrik Do we have any specific reason to hold off merging this? It seems like there may still be underlying issues from upstream that can cause trouble, but the improvements here seem to stand on their own as well for this library.

**Comment by tekktrik** [comment]
I'll defer to @dhalbert but I agree with you.

Also, I didn't fully realize until now that CircuitPython implemented the `for`/`else` combination, cool!  I think this is the first time I've actually seen it in action.

**Comment by dhalbert** [comment]
@FoamyGuy @tekktrik I did not merge it because @BeatArnet still had trouble. But it may be particular samples. So ok, let's merge it, and it will fix the problem at least some of the time.

**Comment by BeatArnet** [comment]
I could solve the problems, using the following code:

```py
# The following code for scanning the I2C addresses is necessary so that the voltage can be measured
BATTERY_MON_ADDRESS = 0x0B  # Address for ESP32-s3 tft Feather

def hack_for_i2c_issue():
    i2c = board.I2C()
    while not i2c.try_lock():
        pass
    running = True
    try:
        while running:
            print("I2C addresses found:", 
                [hex(device_address) for device_address in i2c.scan()]
            )
            # time.sleep(2)
            running = False
        return i2c
    finally:  # unlock the i2c bus when ctrl-c'ing out of the loop
        i2c.unlock()

try:
    print("Searching for battery monitor...")

    i2c = hack_for_i2c_issue()
    battery_monitor = LC709203F(i2c)
```

For me, the problem is solved with the above hack.

**Comment by dhalbert** [comment]
@BeatArnet The code I added is similar to your hack, but only tries three times to wake up the LC709203F. However, if it doesn't succeed, it will throw an exception. So I don't know why your eventually gets read failures. Perhaps the sensor is going back to sleep in a scenario I hadn't envisioned. We could retry reads a few times as well.

**Pull request merged**
